pkg/cmd/template/schema_test.go

```diff
@@ -646,6 +646,198 @@ overriden:
+	opts := cmdtpl.NewOptions()
+	opts.SchemaEnabled = true
+
+	t.Run("when any is true and set on a map and array", func(t *testing.T) {
+		schemaYAML := `#@schema/match data_values=True
+---
+#@schema/type any=True
+foo: ""
+bar:
+#@schema/type any=True
+- 0
+#@schema/type any=True
+baz:
+  a: 1
+`
+		dataValuesYAML := `#@data/values
+---
+foo: ~
+bar: ["", 0, True]
+baz:
+  a: 7
+`
+		templateYAML := `#@ load("@ytt:data", "data")
+---
+foo: #@ data.values.foo
+bar: #@ data.values.bar
+baz: #@ data.values.baz
+`
+		expected := `foo: null
+bar:
+- ""
+- 0
+- true
+baz:
+  a: 7
+`
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("schema.yml", []byte(schemaYAML))),
+			files.MustNewFileFromSource(files.NewBytesSource("dataValues.yml", []byte(dataValuesYAML))),
+			files.MustNewFileFromSource(files.NewBytesSource("template.yml", []byte(templateYAML))),
+		})
+
+		assertSucceeds(t, filesToProcess, expected, opts)
+	})
+	t.Run("when any is true and set on a map but no data values given", func(t *testing.T) {
+		schemaYAML := `#@schema/match data_values=True
+---
+#@schema/type any=True
+foo: ""
+bar:
+#@schema/type any=True
+- 0
+#@schema/type any=True
+baz:
+  a: 1
+`
+		// does a:1 get assigned a 'Type' -> No.
+
+		// should a:1 have a type / what type should it get?
+		// should we build a schema a different way or check in a different way?
+
+		templateYAML := `#@ load("@ytt:data", "data")
+---
+foo: #@ data.values.foo
+bar: #@ data.values.bar
+baz: #@ data.values.baz
+`
+		expected := `foo: ""
+bar: []
+baz:
+  a: 1
+`
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("schema.yml", []byte(schemaYAML))),
+			files.MustNewFileFromSource(files.NewBytesSource("template.yml", []byte(templateYAML))),
+		})
+
+		assertSucceeds(t, filesToProcess, expected, opts)
+	})
+	t.Run("when any is false and set on a map", func(t *testing.T) {
+		schemaYAML := `#@schema/match data_values=True
+---
+#@schema/type any=False
+foo: 0
+`
+		dataValuesYAML := `#@data/values
+---
+foo: 7
+`
+		templateYAML := `#@ load("@ytt:data", "data")
+---
+foo: #@ data.values.foo
+`
+		expected := `foo: 7
+`
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("schema.yml", []byte(schemaYAML))),
+			files.MustNewFileFromSource(files.NewBytesSource("dataValues.yml", []byte(dataValuesYAML))),
+			files.MustNewFileFromSource(files.NewBytesSource("template.yml", []byte(templateYAML))),
+		})
+
+		assertSucceeds(t, filesToProcess, expected, opts)
+	})
+	t.Run("when any is set on maps and arrays with nested dvs and overlay/replace", func(t *testing.T) {
+		schemaYAML := `#@schema/match data_values=True
+---
+#@schema/type any=True
+foo: ""
+bar:
+#@schema/type any=True
+- 0
+#@schema/type any=True
+baz:
+  a: 1
+`
+		dataValuesYAML := `#@data/values
+---
+#@overlay/replace
```

Comment:
This relates to #344. The overlay/replace allows us to get past the overlaying of data values without erroring. Without the overlay/replace, we get overlay errors about expecting a string but found a map.

Reply:
I was able to remove the annotation from the array item, and it did not error. So I'm going to look into this more to see if there is a way to make it possible for maps.

Reply:
The array item does not need the `#@overlay/replace` due to the schema's default data values. When we construct the default dvs for an array, we set the items to be empty: `typedNode.Items = []*yamlmeta.ArrayItem{}`. Since we can overlay onto `nil`, it overlays without the replace annotation. However, arrays are a special case in schemas since they do not have default values.

Reply:
That makes sense why it doesn't work now. IIRC, this is the only place where we require an overlay annotation on the first data values document, so it's a divergent experience. Curious if you have thoughts on whether it would be possible to tackle here or if it makes sense to separate out?